src/dags/dag_tree.py

```diff
@@ -0,0 +1,323 @@
+) -> FlatFunctionDict:
+    flat_functions = _flatten_str_dict(functions)
+    flat_input_structure = _flatten_str_dict(input_structure)
+
+    _check_functions_and_input_overlap(
+        flat_functions, flat_input_structure, name_clashes
+    )
+
+    for name, function in flat_functions.items():
+        namespace = _qualified_name_delimiter.join(
+            name.split(_qualified_name_delimiter)[:-1]
+        )
+
+        renamed = rename_arguments(
+            function,
+            mapper=_create_parameter_name_mapper(
+                flat_functions,
+                flat_input_structure,
+                namespace,
+                function,
+            ),
+        )
+
+        flat_functions[name] = renamed
+
+    return flat_functions
+
+
+def _check_functions_and_input_overlap(
+    flat_functions: FlatFunctionDict,
+    input_structure: FlatInputStructureDict,
+    name_clashes: Literal["raise", "warn", "ignore"],
+) -> None:
+    overlap = set(flat_functions.keys()) & set(input_structure.keys())
+
+    if overlap:
+        message = f"These names are both functions and inputs: {', '.join(overlap)}."
+
+        if name_clashes == "raise":
+            raise ValueError(message)
+        elif name_clashes == "warn":
+            warnings.warn(message)
+        elif name_clashes == "ignore":
+            pass
+
+
+def _create_parameter_name_mapper(
+    flat_functions: FlatFunctionDict,
+    flat_input_structure: FlatInputStructureDict,
+    namespace: str,
+    function: Callable,
+) -> dict[str, str]:
+    return {
+        old_name: _map_parameter(
+            flat_functions, flat_input_structure, namespace, old_name
+        )
+        for old_name in _get_free_arguments(function)
+    }
+
+
+def _map_parameter(
+    flat_functions: FlatFunctionDict,
+    flat_input_structure: FlatInputStructureDict,
+    namespace: str,
+    parameter_name: str,
+) -> str:
+    # Parameter name is definitely a qualified name
+    if _is_qualified_name(parameter_name):
+        return parameter_name
+
+    # (1.1) Look for function in the current namespace
+    namespaced_parameter = (
+        f"{namespace}__{parameter_name}" if namespace else parameter_name
+    )
+    if namespaced_parameter in flat_functions:
+        return namespaced_parameter
+
+    # (1.2) Look for input in the current namespace
+    if namespaced_parameter in flat_input_structure:
+        return namespaced_parameter
+
+    # (2.1) Look for function in the top level
+    if parameter_name in flat_functions:
+        return parameter_name
+
+    # (2.2) Look for input in the top level
+    if parameter_name in flat_input_structure:
+        return parameter_name
+
+    # (3) Raise error
+    raise ValueError(f"Cannot resolve parameter {parameter_name}")
+
+
+def create_input_structure_tree(
+    functions: NestedFunctionDict,
+    targets: Optional[NestedTargetDict] = None,
+    level_of_inputs: TopOrNamespace = "namespace",
```

Comment:
```suggestion
    namespace_of_inputs: GlobalOrLocal = "local",
```


with:

```py
GlobalOrLocal = Literal["global", "local"]
```

This occurred to me the other day -- why don't we just borrow Python terminology rather than re-inventing the wheel?

Reply:
Thanks for the quick implementation! I'd also be happy to have a discussion first, I do have wild ideas at times and I do oversee things... 

Any reason you prefer `level_of_inputs` over `namespace_of_inputs` ?

@janosg, thoughts on your behalf on the name and options?

Reply:
> Any reason you prefer level_of_inputs over namespace_of_inputs ?

No, I just missed that part.